runtime-sdk/modules/evm/src/precompile/subcall.rs

```diff
@@ -0,0 +1,521 @@
+use ethabi::{ParamType, Token};
+use evm::{
+    executor::stack::{PrecompileFailure, PrecompileHandle, PrecompileOutput},
+    ExitError, ExitSucceed,
+};
+
+use crate::backend::EVMBackendExt;
+use oasis_runtime_sdk::{
+    module::CallResult, modules::core::Error, subcall, types::transaction::CallerAddress,
+};
+
+use super::{record_linear_cost, PrecompileResult};
+
+/// A subcall validator which prevents any subcalls from re-entering the EVM module.
+struct ForbidReentrancy;
+
+impl subcall::Validator for ForbidReentrancy {
+    fn validate(&self, info: &subcall::SubcallInfo) -> Result<(), Error> {
+        if info.method.starts_with("evm.") {
+            return Err(Error::Forbidden);
+        }
+        Ok(())
+    }
+}
+
+const SUBCALL_BASE_COST: u64 = 10;
+const SUBCALL_WORD_COST: u64 = 1;
+
+pub(super) fn call_subcall<B: EVMBackendExt>(
+    handle: &mut impl PrecompileHandle,
+    backend: &B,
+) -> PrecompileResult {
+    record_linear_cost(
+        handle,
+        handle.input().len() as u64,
+        SUBCALL_BASE_COST,
+        SUBCALL_WORD_COST,
+    )?;
+
+    // Ensure that the precompile is called using a regular call so the caller is actually the
+    // address of the calling contract.
```

Comment:
I think it would be useful if this mentions "delegatecall", at least to me, it was not immediately obvious why this is needed.